src/main/flight/imu.c

```diff
@@ -215,51 +213,79 @@ static float invSqrt(float x)
-        // Note that Earth frame X is pointing north and sin/cos argument is anticlockwise
+    // *** Calculate and correct IMU heading from GPS values once min groundspeed exceeded ***
+    if (gpsGroundspeedGain > 0.0f) {
+        // Compute GPS heading vector in earth frame (ef) from scalar GPS courseOverGround
+        // Earth frame X is pointing north and sin/cos argument is anticlockwise
         const fpVector2_t cog_ef = {.x = cos_approx(-courseOverGround), .y = sin_approx(-courseOverGround)};
-#define THRUST_COG 1
-#if THRUST_COG
-        const fpVector2_t heading_ef = {.x = rMat[X][Z], .y = rMat[Y][Z]};  // body Z axis (up) - direction of thrust vector
-#else
-        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};  // body X axis. Projected vector magnitude is reduced as pitch increases
-#endif
-        // cross product = 1 * |heading| * sin(angle) (magnitude of Z vector in 3D)
-        // order operands so that rotation is in direction of zero error
+
+        // Compute IMU / craft heading vector and magnitude in body X axis
+        // Important:  vector magnitude is reduced as pitch increases
+        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};
+        const float headingMagnitude = vector2Mag(&heading_ef);
+
+        // cross (product) = heading * cogHeading * sin error angle, ie, reflects error angle of quad X axis vs GPS groundspeed
+        // cross value sign depends on the rotation change from input to output vectors by right hand rule
+        // operand order is such that rotation is in direction of zero error
+        // abs value of cross is zero when parallel, max of 1.0 at 90 degree error
+        // cross value determines ez_ef when error is less than 90 degrees
+        // a zero cross value at 180 degree error would prevent error correction
         const float cross = vector2Cross(&heading_ef, &cog_ef);
-        // dot product, 1 * |heading| * cos(angle)
+
+        // dot product = heading * cogHeading * cos(angle)
+        // max when aligned, zero at 90 degrees of vector difference, negative for any error > 90 degrees
         const float dot = vector2Dot(&heading_ef, &cog_ef);
-        // use cross product / sin(angle) when error < 90deg (cos > 0),
-        //   |heading| if error is larger (cos < 0)
-        const float heading_mag = vector2Mag(&heading_ef);
-        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * heading_mag;
-#if THRUST_COG
-        // increase gain for small tilt (just heuristic; sqrt is cheap on F4+)
-        ez_ef /= sqrtf(heading_mag);
-#endif
-        ez_ef *= cogYawGain;          // apply gain parameter
+
+
+        // ez_ef is the error value used to correct the heading via an IIR like smoothing function
+        // for error angles less than 90 degrees (dot > 0), use cross value to compute ez_ef
+        // above 90 degrees error, use heading vector magnitude, but give it sign from cross product
```

Comment:
cross = |heading| * |cogHeading| * sin(error_angle)

error < 90 deg: cross product is increasing with error (-> more correction). Also, it is almost linear on small `error_angle`, which may help in error minimization code.
 
error > 90 deg: sin(angle) start decreasing (and so cross product), so bigger input would lead to smaller correction (and no correction at 180deg). 

Above 90 deg, sin(angle) is replaced with sin(90 deg) == 1 - correction at 90 deg. 
Same for error over -90deg, just use sin(-90) == -1.
This is sign function.

Cross product was |heading| * |cogHeading| * sin(error_angle), and was computed using vector multiplication, no explicit sin necessary. 
Now we want |heading| * |cogHeading| * sign(error_angle). We must compute it explicitly.  |cogHeading| == 1, so only |heading| * sign(error_angle) remains

Reply:
My comment was about use of `headingMagnitude` above 90 degrees of heading error.

Your comment that the pitch angle of the quad would reduce `headingMagnitude` suggested that `headingMagnitude` reflected the magnitude of the X axis of the quad heading vector in the earth frame.  That vector would be attenuated in the earth frame to a value below 1 if the quad was pitched forward (or backward) in relation to the earth frame.

And from what I could see, the value of `cross` as the angle approached 90 degrees would always be 1.0, since its calculation only considered the angles, not the magnitudes, of those vectors.

Hence my concern was that we could approach, and then exceed, a 90 degree angle between the two vectors, and that `cross` would approach 1.0 as we got closer to 90 degrees, regardless of pitch angle, and then suddenly we might return a magnitude value below 1.0 due to the influence of pitch angle on heading magnitude the instant we exceed 90 degrees.

Reply:
Cross product is proportional to size of input vectors too. |A| * |B| * sin(angle), where |A| is length/magnitude of A

Reply:
So if the craft is pitched forward, say 45 degrees, with no roll, and the IMU earth frame heading angle happens to be exactly the same as the X axis heading in the earth frame, ie they are perfectly aligned, is `cross` no longer 1.0?  

And if that's the case, cross isn't a very good indicator of alignment; wouldn't we just want the estimate of the error in angle terms alone?  After all, to correct the heading angle, we just want to know about whether the craft X axis heading is parallel with the IMU estimate, don't we?  Why would we care about magnitude in the context of heading angle accuracy?

Reply:
I guess I mean that if we have a `heading_ef` vector, can we not just compare the 'angle' it represents to the 'angle' represented by the IMU heading, both in the earth frame, and get a unitary value between 0 and 1 that represents the 'error in the heading angle', and nothing else?

Otherwise if the indicator of heading 'angle error' returns 0.7, we don't know if we have an error in angle that needs correcting, or if we have perfect angular alignment but happen to be pitched forward at the same time?

Reply:
It is possible to normalize `heading_ef`, then error will be dependent only on angle. But then you have to handle case when quad is pointing up - normalization will lead to division by zero.

Without normalization, |heading_ef| is taken as 'heading information quality':
|heading_ef| ==1 - quad is level, we have very good idea about direction
|heading_ef| ==0 - quad is pointing directly up, heading makes not sense (in 'which NSWE direction is front pointing' sense).
And we save one square root and division as a bonus.

> Otherwise if the indicator of heading 'angle error' returns 0.7, we don't know if we have an error in angle that needs correcting, or if we have perfect angular alignment but happen to be pitched forward at the same time?

Only important thing is that (nonzero) error is pointing in correct direction and it diminishing to 0 when orientation is correct. But error==0 means either that there is not error OR that we have no idea about error magnitude.

'angle error' == .7 means that hearing is pretty badly misaligned and .7 being > 0 tells IMU which that it should increase heading estimate.

Reply:
I have implemented a normalisation, I think it would return zero instead of making a divide by zero; this may glitch but should be brief and would not usually happen when the quad is being controlled based on attitude.